core/src/test/java/com/linecorp/armeria/server/ServerChildChannelPipelineCustomizerTest.java

```diff
@@ -0,0 +1,120 @@
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.internal.common.util.SslContextUtil;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.traffic.ChannelTrafficShapingHandler;
+
+class ServerChildChannelPipelineCustomizerTest {
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.http(0);
+            sb.https(0);
+            sb.tlsSelfSigned();
+            sb.childChannelPipelineCustomizer(
+                    pipeline ->  pipeline.addFirst(new ChannelTrafficShapingHandler(1024, 0)));
+            sb.service("/", new AbstractHttpService() {
+                @Override
+                protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req)
+                        throws Exception {
+                    return HttpResponse.of(
+                        ResponseHeaders.of(HttpStatus.OK, "header1", Strings.repeat("a", 2048)));
+                }
+            });
+        }
+    };
+
+    static final SSLEngine sslEngine =  SslContextUtil.createSslContext(
+            SslContextBuilder::forClient,
+            /* forceHttp1 */ true,
+            /* tlsAllowUnsafeCiphers */ false,
+            ImmutableList.of()).newEngine(ByteBufAllocator.DEFAULT);
+
+    @RegisterExtension
+    static ServerExtension server2 = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.http(8080);
+            sb.childChannelPipelineCustomizer(
+                    pipeline ->  pipeline.addFirst(new SslHandler(sslEngine)));
+            sb.service("/", (ctx, req) -> HttpResponse.of("content"));
+        }
+    };
+
+    @EnumSource(value = SessionProtocol.class, names = {"H1", "H1C"})
+    @ParameterizedTest
+    void testResponseTimeout(SessionProtocol protocol) {
+        final RequestHeaders requestHeaders = RequestHeaders.of(HttpMethod.GET, "/");
+        final ClientFactory clientFactory = ClientFactory.builder()
+                                                         .tlsNoVerify()
+                                                         .build();
+
+        // using h1 or h1c since http2 compresses headers
+        assertThatThrownBy(() -> WebClient.builder(server.uri(protocol))
+                                          .responseTimeoutMillis(1000)
+                                          .factory(clientFactory)
+                                          .build()
+                                          .blocking()
+                                          .execute(requestHeaders, "content"))
+                .isInstanceOf(ResponseTimeoutException.class);
```

Comment:
This should actually succeed?

Reply:
Yes, this should actually succeed.

By setting Channel Pipeline Customizer like the following, the response from the server becomes to be delayed and this exception occurs.
```java
sb.childChannelPipelineCustomizer(
                    pipeline -> pipeline.addFirst(new ChannelTrafficShapingHandler(1024, 0)));
```